### PR TITLE
LevelUX: Stop capstone autoplay

### PR DESCRIPTION
**Pull request by codeluggage**
# Feature

## Why are these changes necessary? 

Capstone levels (all game dev levels) automatically update the goals and the level session when the student's code changes. This is no longer the wanted behavior for capstone levels. Instead, the new `Update` button should update the goals and the level session. The `Next` button should trigger the completion of the level (victory dialogue or progression to next cinematic), when the goals of the current capstone stage have been completed at least once. 

## What has changed?

The CastButtonView has several updates to how the buttons work, with a new call to recompile the code in the SpellView. 

The SpellView now excludes automatic updates for capstone levels, and handles recompiles when requested.

### CODE REVIEW

File: ozaria/site/views/play/level/tome/CastButtonView.coffee [hunk 1]
```diff
@@ -75,21 +73,10 @@ module.exports = class CastButtonView extends CocoView
...
-    @updateReplayability()
-
-  onClickGameDevPlayButton: ->
-    Backbone.Mediator.publish 'tome:manual-cast', {realTime: true}
-
-  onDoneButtonClick: (e) ->
-    return if @options.level.hasLocalChanges()  # Don't award achievements when beating level changed in level editor
+  onNextButtonClick: (e) ->
     @options.session.recordScores @world?.scores, @options.level
     Backbone.Mediator.publish 'level:show-victory', { showModal: true, manual: true }
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/tome/CastButtonView.coffee]: You will need to pass `capstoneInProgress: true/false` here for the victory modal as we discussed offline.

File: ozaria/site/views/play/level/tome/SpellView.coffee [hunk 2]
```diff
@@ -564,6 +565,10 @@ module.exports = class SpellView extends CocoView
     @createToolbarView()
     @updateHTML create: true if @options.level.isType('web-dev')
 
+  onRecompile: ->
+    @recompile
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/tome/SpellView.coffee]: Left by mistake?

File: ozaria/site/views/play/level/LevelPlaybackView.coffee [hunk 3]
```diff
@@ -303,6 +303,9 @@ module.exports = class LevelPlaybackView extends CocoView
   onTogglePlay: (e) ->
     e?.preventDefault?()
     return if @shouldIgnore()
+    playing = store.state.game.playing
+    if not playing
+      Backbone.Mediator.publish('tome:manual-cast', {realTime: true})
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/LevelPlaybackView.coffee]: Does this piece of code need to execute for all levels, or just capstone?
Also, I think the UI of the play button/the whole LevelPlayBackView is different for the capstone, is that going to be handled in a separate PR?

File: ozaria/site/views/play/level/tome/SpellView.coffee [hunk 4]
```diff
@@ -733,13 +738,14 @@ module.exports = class SpellView extends CocoView
...
     onAnyChange = [
-      _.debounce @updateAether, if @options.level.isType('game-dev') then 10 else 500
       _.debounce @notifyEditingEnded, 1000
       _.throttle @notifyEditingBegan, 250
       _.throttle @notifySpellChanged, 300
       _.throttle @updateLines, 500
       _.throttle @hideProblemAlert, 500
     ]
+    unless @options.level.isType('game-dev')
+      onAnyChange.push(_.debounce(@updateAether, 500))
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/tome/SpellView.coffee]: Why is this not required for game-dev/capstone? Also do we want to do this for all game-dev levels or just capstone, or probably its the same thing(i.e. only capstones can be game-dev)? I am not sure, but might be worth checking.

File: ozaria/site/views/play/level/tome/SpellView.coffee [hunk 5]
```diff
@@ -960,11 +966,7 @@ module.exports = class SpellView extends CocoView
     beginningOfLine = not currentLine.substr(0, cursorPosition.column).trim().length  # uncommenting code, for example
     incompleteThis = /^(s|se|sel|self|t|th|thi|this|g|ga|gam|game|h|he|her|hero)$/.test currentLine.trim()
     #console.log "finished=#{valid and (endOfLine or beginningOfLine) and not incompleteThis}", valid, endOfLine, beginningOfLine, incompleteThis, cursorPosition, currentLine.length, aether, new Date() - 0, currentLine
-    if not incompleteThis and @options.level.isType('game-dev')
-      # TODO: Improve gamedev autocast speed
-      @spell.transpile @getSource()
-      @cast(false, false, true)
-    else if (endOfLine or beginningOfLine) and not incompleteThis
+    if (endOfLine or beginningOfLine) and not incompleteThis
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/tome/SpellView.coffee]: It seems it will now run for game-dev levels also, is that intended?

**Comment by shubhi1092** [comment]
Left some questions!

File: ozaria/site/views/play/level/tome/SpellView.coffee (hunk 2 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/tome/SpellView.coffee]: This is required because the updating of the spell has been disabled as part of this PR. If we only send the event, it won't update everything it needs to. To recompile and send the event was what covered all the use cases.

File: ozaria/site/views/play/level/LevelPlaybackView.coffee (hunk 3 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/LevelPlaybackView.coffee]: For other levels, it would usually go through the "Run" button. As this area is swapped out with the UX for the scrubber area, this can move there. It will be handled in a separate PR.

File: ozaria/site/views/play/level/tome/SpellView.coffee (hunk 4 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/tome/SpellView.coffee]: I don't know of game dev levels that are not capstone now, but in case we add that in the future I will add an additional check. 

This change means that the level is not constantly updated by the code changing, which is the purpose of this PR.

File: ozaria/site/views/play/level/tome/SpellView.coffee (hunk 5 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/tome/SpellView.coffee]: Preloading is OK, but automatically casting is not. I believe that without this, neither the if nor the else gets called for game dev levels, which can block the full loading of the level. 

Now that we want to be selective about updating the level from writing code, it should behave quite similarly for game dev and non game dev levels.

File: ozaria/site/views/play/level/LevelPlaybackView.coffee (hunk 3 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/LevelPlaybackView.coffee]: The state of the buttons and the cast'ing is tricky between 4 states: 
- regular levels in coco
- game dev levels in coco
- regular levels in ozaria
- game dev (or capstone) levels in ozaria

This is handled in the next PR for the play button.
